app/src/chainlit.py

```diff
@@ -1,32 +1,62 @@
 import logging
+import os
+from urllib.parse import parse_qs, urlparse
 
 import chainlit as cl
-import src.adapters.db as db
+from src import chat_engine
 from src.format import format_guru_cards
-from src.generate import generate
 from src.login import require_login
-from src.retrieve import retrieve
-from src.shared import get_embedding_model
 
 logger = logging.getLogger(__name__)
 
 require_login()
 
 
-@cl.on_message
-async def main(message: cl.Message) -> None:
-    logger.info(f"Received: {message.content!r}")
+@cl.on_chat_start
+async def start() -> None:
+    engine_id = engine_url_query_value()
+    logger.info("Engine ID: %s", engine_id)
+    engine = chat_engine.create_engine(engine_id)
+    if not engine:
+        await cl.Message(
+            author="backend",
+            metadata={"engine": engine_id},
+            content=f"Available engines: {chat_engine.available_engines()}",
+        ).send()
+        return
+
+    cl.user_session.set("chat_engine", engine)
+    await cl.Message(
+        author="backend",
+        metadata={"engine": engine_id},
+        content=f"Chat engine started: {engine.name}",
+    ).send()
 
-    with db.PostgresDBClient().get_session() as db_session:
-        chunks = retrieve(
-            db_session,
-            get_embedding_model(),
-            message.content,
-        )
 
-    response = generate(message.content, context=chunks)
-    formatted_guru_cards = format_guru_cards(chunks)
+def engine_url_query_value() -> str:
+    url = cl.user_session.get("http_referer")
+    logger.debug("Referer URL: %s", url)
 
-    await cl.Message(
-        content=response + formatted_guru_cards,
-    ).send()
+    # Using this suggestion: https://github.com/Chainlit/chainlit/issues/144#issuecomment-2227543547
+    parsed_url = urlparse(url)
+    qs = parse_qs(parsed_url.query)
+    return qs.get("engine", [os.environ.get("CHAT_ENGINE", "default_engine")])[0]
+
+
+@cl.on_message
+async def on_message(message: cl.Message) -> None:
+    logger.info("Received: %r", message.content)
+
+    engine: chat_engine.ChatEngineInterface = cl.user_session.get("chat_engine")
+    try:
+        result = engine.on_message(question=message.content)
+        msg_content = result.response + format_guru_cards(result.chunks)
```

Comment:
Clarifying my earlier post: I like separating view + model, which in your original implementation was happening by defining two different functions in GuruBaseEngine -- I would keep those as is, but make GuruBaseEngine.on_message or some other function responsible for calling both. So the control flow looks like:

```python
# chainlit.py
def on_message():
    response = engine.on_message(question=message.content)
    ...
```

```python
# chat_engine.py
class GuruBaseEngine:
    def on_message(message: str) -> str:
        result = process(message)
        return format(result)
        
    def process(message: str) -> type TBD by GuruBaseEngine
        ... # Controller
    
    def format(type TBD) -> str:
        ... # View
```

Reply:
Yes, that's how I understood your original message. I actually don't want any UI/text formatting in the chat engines -- I see them as backend implementations. Any formatting should be done by a separate entity. So it's good that you convinced me to remove chainlit imports from `chat_engine.py`, which convinced me to restore `format_guru_cards()` back in `chainlit.py`. 

I think a chat engine can specify the output formatting as a configuration value, and some other (view-centric) entity can perform that formatting.